llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -2041,8 +2041,13 @@ bool AArch64TargetLowering::shouldExpandPartialReductionIntrinsic(
     return true;
 
   EVT VT = EVT::getEVT(I->getType());
-  return VT != MVT::nxv4i64 && VT != MVT::nxv4i32 && VT != MVT::nxv2i64 &&
-         VT != MVT::v4i64 && VT != MVT::v4i32 && VT != MVT::v2i32;
+  auto Op1 = I->getOperand(1);
+  EVT Op1VT = EVT::getEVT(Op1->getType());
+  if (Op1VT.getVectorElementType() == VT.getVectorElementType() &&
+      (VT.getVectorElementCount() * 4 == Op1VT.getVectorElementCount() ||
+       VT.getVectorElementCount() * 2 == Op1VT.getVectorElementCount()))
```

Comment:
The previous check limited this to specific types, this only looks at the element counts, but presumably not all of these can be mapped to instructions. What happens int those cases?

Reply:
It looks like it should fallback to `DAG.getPartialReduceAdd()` (in `performIntrinsicCombine()`), but `tryLowerPartialReductionToWideAdd()` does not seem to check if there is a wide add instruction available.

Reply:
I did this with the view of implementing type legalisation I think. For example, we do want a partial reduction from `nxv4i16` to `nxv2i32` (after the input has been extended) to be possible right? And I thought that all that would be needed to be done in checking the wide add instruction is available is checking whether SVE was enabled, or is there something I'm missing?

Reply:
I think you may be right, it doesn't seem like other intrinsics like dot product have legalisation. So do you think that all the tests required are already in `sve-partial-reduce-wide-add.ll` ? `(`nxv4i32`->nxv2i64`, `nxv8i16`->`nxv4i32`, `nxv16i8`->`nxv8i16`)

Reply:
This is now resolved with an additional check in the `tryLowerPartialReductionToWideAdd` function.

Reply:
Cool, thanks!